feat(core): ESM support in policy generation

boneskull opened the pull request:
This adds ESM support to the policy generator.

## TODO

- [x] add ~~scenarios~~ test project(s). only policy gen will be tested since the runtime cannot execute ESM
- [x] handle CJS/ESM disagreements about allowed global variable names

boneskull committed:
feat(core): ESM support in policy generation

This adds ESM support to the policy generator.

naugtur commented on packages/core/src/generatePolicy.js:
```diff
@@ -251,7 +255,14 @@ function createModuleInspector(opts) {
+      'exports',
+      'arguments',
+      'import',
+      'export',
```
this seems incorrect. I can definitely create a function named `import` or use a global named `import` in a cjs module.

boneskull replied:
Indeed!  This will have to make a distinction between CJS and ESM earlier; this part of the code will need to know what module system is in play.  I don't know any way to do that other than to first check file extension, and if it's `.js`, read `package.json` (or use a lib to do something like this).

boneskull replied:
Also, I think we shouldn't do that in `core` at all; the consuming module (lavamoat, endomoat, webpack, etc) will need to tell it (via callback or otherwise) what module system to use.

(why wasn't `import` a reserved keyword? urgh)

boneskull replied:
since this wants a `LavamoatModuleRecord`, it may make sense to stuff the information in there, wouldn't it?

now I am wondering why `ast` is its own param and not `moduleRecord.ast`

boneskull replied:
@naugtur I'm going to sit on this until you respond; I don't want to implement this in e.g., `lavamoat-node` or elsewhere if it's not the right direction.

naugtur replied:
Let's discuss this synchronously.
If we don't know what's esm or cjs in here, how do we inspectESMImports?
Endo doesn't support dynamic imports anyway, so maybe that changes our assumptions?
Does the AST specify the difference between dynamic import and a function named import?

boneskull committed:
fix(core): detect ESM/CJS imports properly

Babel can deduce the source type (CJS/ESM) of the code it's parsed; this lives in the `program.sourceType` nested property of of the `File` object it `parse()` returns. We use this to determine the symbols in the module scope, which should be excluded from the list of globals the inspector finds.

boneskull committed:
fix(lavamoat): update for ESM support in policy gen

naugtur commented on packages/core/src/generatePolicy.js:
```diff
@@ -162,7 +174,7 @@ function createModuleInspector(opts) {
       parse(moduleRecord.content, {
         // esm support
-        sourceType: 'module',
+        sourceType: 'unambiguous',
```
what does this mean? if it's not ambiguous, why doesn't it have a different value?

boneskull replied:
it means, "babel, please determine which module type this is".  I am not sure why it's called "unambiguous", though; it's a confusing property.
